sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusReceiverAsyncClientIntegrationTest.java

```diff
@@ -63,26 +64,41 @@
     @Override
     protected void beforeTest() {
         sessionId = UUID.randomUUID().toString();
+        messagesDeferredPending.set(new ArrayList<>());
     }
 
     @Override
     protected void afterTest() {
         sharedBuilder =  null;
         final int pending = messagesPending.get();
-        if (pending < 1) {
+        final int pendingDeferred = messagesDeferredPending.get().size();
+        if (pending < 1 && pendingDeferred < 1) {
             dispose(receiver, sender, receiveAndDeleteReceiver);
             return;
         }
 
         // In the case that this test failed... we're going to drain the queue or subscription.
         try {
-            receiveAndDeleteReceiver.receive()
-                .map(message -> {
-                    logger.info("Message received: {}", message.getMessage().getSequenceNumber());
-                    return message;
-                })
-                .timeout(Duration.ofSeconds(15), Mono.empty())
-                .blockLast();
+            if (pending > 0) {
+                receiveAndDeleteReceiver.receive()
+                    .map(message -> {
+                        logger.info("Message received: {}", message.getMessage().getSequenceNumber());
+                        return message;
+                    })
+                    .timeout(Duration.ofSeconds(15), Mono.empty())
+                    .blockLast();
+            }
+            if (pendingDeferred > 0) {
```

Comment:
There should be a try-catch finally around each block. Even if the `pending > 0` portion throws an exception, I would want `messagesDeferrenedPending` to still run.